File: app/src/main/java/tech/alvarez/numbers/adapters/SearchChannelsAdapter.java
package tech.alvarez.numbers.adapters;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;

import java.util.ArrayList;
import java.util.List;

import tech.alvarez.numbers.R;
import tech.alvarez.numbers.models.youtube.search.ItemSearchResponse;
import tech.alvarez.numbers.utils.Database;

/**
 * Created by Daniel Alvarez on 8/7/16.
 * Copyright © 2016 Alvarez.tech. All rights reserved.
 */
public class SearchChannelsAdapter extends RecyclerView.Adapter<SearchChannelsAdapter.ViewHolder> {

    private Context context;
    private List<ItemSearchResponse> dataset;
    private OnItemClickListener onItemClickListener;

    public SearchChannelsAdapter(Context context, OnItemClickListener onItemClickListener) {
        this.context = context;
        this.dataset = new ArrayList<>();
        this.onItemClickListener = onItemClickListener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_channel, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        ItemSearchResponse item = dataset.get(position);
        holder.titleTextView.setText(item.getSnippet().getTitle());
        holder.descTextView.setText(item.getSnippet().getDescription());
        String url = "";

        if (item.getSnippet().getThumbnails().getDefaultThumbnail() != null) {
            url = item.getSnippet().getThumbnails().getDefaultThumbnail().getUrl();
        }
        if (Database.inTheDatabase(item.getSnippet().getChannelId())) {
            holder.addButton.setVisibility(View.GONE);
        } else {
            holder.addButton.setVisibility(View.VISIBLE);
        }

        Glide.with(context).load(url)
                .crossFade()
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .into(holder.photoImageView);

        holder.setOnItemClickListener(item, onItemClickListener);
    }

    @Override
    public int getItemCount() {
        return dataset.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        private ImageView photoImageView;
        private TextView titleTextView;
        private TextView descTextView;
        private Button addButton;

        public ViewHolder(View itemView) {
            super(itemView);
            photoImageView = (ImageView) itemView.findViewById(R.id.photoImageView);
            titleTextView = (TextView) itemView.findViewById(R.id.titleTextView);
            descTextView = (TextView) itemView.findViewById(R.id.descTextView);
            addButton = (Button) itemView.findViewById(R.id.addButton);
        }

        public void setOnItemClickListener(final ItemSearchResponse channel, final OnItemClickListener onItemClickListener) {
            addButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    onItemClickListener.onItemClick(channel);
                }
            });
        }
    }

    public List<ItemSearchResponse> getDataset() {
        return dataset;
    }

    public void setDataset(List<ItemSearchResponse> dataset) {
        this.dataset = dataset;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/tech/alvarez/numbers/adapters/SearchChannelsAdapter.kt
package tech.alvarez.numbers.adapters

import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import tech.alvarez.numbers.R
import tech.alvarez.numbers.model.ItemSearchResponse
import tech.alvarez.numbers.util.inflate

interface OnItemClickListener {
    fun onItemClick(channel: ItemSearchResponse?)
}

class SearchChannelsAdapter(onItemClickListener: OnItemClickListener) :
    RecyclerView.Adapter<SearchChannelsAdapter.ChannelHolder>() {
    private var dataset: List<ItemSearchResponse> = listOf()
    private val onItemClickListener: OnItemClickListener = onItemClickListener

    init {
//        mDb = AppDatabase.getDatabase(context.applicationContext)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ChannelHolder {
        val inflatedView = parent.inflate(R.layout.item_channel, false)
        return ChannelHolder(inflatedView)
    }

    override fun onBindViewHolder(
        holder: ChannelHolder,
        position: Int
    ) {
        val item: ItemSearchResponse = dataset[position]
        holder.bindData(item)
        holder.setOnItemClickListener(item, onItemClickListener)
    }

    override fun getItemCount() = dataset.size

    class ChannelHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val photoImageView: ImageView =
            itemView.findViewById<View>(R.id.photoImageView) as ImageView
        private val titleTextView: TextView =
            itemView.findViewById<View>(R.id.titleTextView) as TextView
        private val descTextView: TextView =
            itemView.findViewById<View>(R.id.descTextView) as TextView
        private val addButton: Button = itemView.findViewById<View>(R.id.addButton) as Button

        fun setOnItemClickListener(
            channel: ItemSearchResponse?,
            onItemClickListener: OnItemClickListener
        ) {
            addButton.setOnClickListener { onItemClickListener.onItemClick(channel) }
        }

        fun bindData(item: ItemSearchResponse) {
            titleTextView.text = item.snippet?.title
            descTextView.text = item.snippet?.description
            var url = ""
            if (item.snippet?.thumbnails?.defaultThumbnail != null) {
                url = item.snippet?.thumbnails?.defaultThumbnail?.url ?: ""
            }
            Glide.with(itemView.context).load(url)
                .crossFade()
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .into(photoImageView)
        }
    }

    fun setDataset(dataset: List<ItemSearchResponse>) {
        this.dataset = dataset
        notifyDataSetChanged()
    }
}
